zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/Follower.java

```diff
@@ -150,13 +157,95 @@ void followLeader() throws InterruptedException {
         }
     }
 
+    // Visible for testing
+    protected void snapPing(QuorumPacket qp) throws IOException {
+        int peerSnapPingVersion = -1;
+        long snapPingId = -1;
+        int snapCode = -1;
+        try {
+            ByteArrayInputStream bis = new ByteArrayInputStream(qp.getData());
+            DataInputStream dis = new DataInputStream(bis);
+            peerSnapPingVersion = dis.readInt();
+            snapPingId = dis.readLong();
+            snapCode = dis.readInt();
+        } catch (IOException e) {
+            LOG.error("Error while parsing SnapPing packet", e);
+            return;
+        }
+
+        if (peerSnapPingVersion != SnapPingManager.SNAP_PING_VERSION) {
+            LOG.warn("The SnapPing version on leader is {} which is not "
+                    + "compatible with ours {}, will skip", peerSnapPingVersion,
+                    SnapPingManager.SNAP_PING_VERSION);
+            if (fzk.syncProcessor.isOnlySnapWhenSafetyIsThreatened()) {
+                LOG.info("SnapPing version incompatible, start self snapshot");
+                fzk.syncProcessor.setOnlySnapWhenSafetyIsThreatened(false);
+            }
+            return;
+        }
+
+        if (snapCode == SnapPingCode.CANCEL.ordinal()) {
+            if (fzk.syncProcessor.isOnlySnapWhenSafetyIsThreatened()) {
+                LOG.info("Snapshot schedule cancelled by leader, start self snapshot");
+                fzk.syncProcessor.setOnlySnapWhenSafetyIsThreatened(false);
```

Comment:
did we consider that the CANCEL packet could potentially lost so a server in schedule snap mode will never be snapping again? Is there any built in defense mechanism for that case (didn't read all part of code yet) to make sure a server will not end up in not snapping state?